OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,338 @@
+                {
+                    if (!IsPressed(combo))
+                    {
+                        ((ModInputCombination)combo).SetPressed(false);
+                        toRemove.Add(combo);
+                    }
+                }
+                foreach (var combo in toRemove)
+                {
+                    _singlePressed.Remove(combo);
+                }
+            }
+            foreach (var key in ((ModInputCombination)combination).GetSingles())
+            {
+                if (UnityEngine.Input.GetKey(key) && !ShouldIgnore(key))
+                {
+                    _singlePressed.Add(combination);
+                    ((ModInputCombination)combination).SetPressed();
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public bool IsPressed(IModInputCombination combination)
+        {
+            return IsPressedSingle(combination) || IsPressedExact(combination);
+        }
+
+        public bool IsNewlyPressed(IModInputCombination combination, bool keep = false)
+        {
+            return IsPressed(combination) && combination.IsFirst(keep);
+        }
+
+        public bool WasTapped(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && (combination.PressDuration < TapDuration) 
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleased(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && combination.IsFirst(keep);
+        }
+
+        private KeyCode StringToKeyCode(string key)
+        {
+            key = key.Trim();
+            if (key.Contains(XboxPrefix))
+            {
+                var xboxKey = key.Substring(XboxPrefix.Length);
+                var xboxCode = (XboxButton)Enum.Parse(typeof(XboxButton), xboxKey, true);
+                if (Enum.IsDefined(typeof(XboxButton), xboxCode))
+                {
+                    return InputTranslator.GetKeyCode(xboxCode, false);
+                }
+                return KeyCode.None;
+            }
+            var changedKey = key;
+            if (key == "control" || key == "ctrl")
+            {
+                return KeyCode.LeftControl;
+            }
+            if (key == "shift")
+            {
+                return KeyCode.LeftShift;
+            }
+            if (key == "alt")
+            {
+                return KeyCode.LeftAlt;
+            }
+            var code = (KeyCode)Enum.Parse(typeof(KeyCode), changedKey, true);
+            if (Enum.IsDefined(typeof(KeyCode), code))
+            {
+                return code;
+            }
+            return KeyCode.None;
+        }
+
+        private long ParseCombination(string combo, bool forRemoval = false)
```

Comment:
This method is a bit complicated to read. I think it would be nice if you extract the contents into smaller methods with intuitive names.

Reply:
done, also it's StringToHash now